whois.1
.TH "WHOIS" "1" "2019-12-30" "Marco d'Itri" "Debian GNU/Linux"
.SH NAME
whois \- client for the whois directory service
.SH SYNOPSIS
.B whois
[
.RB {\~ \-h \~|\~ \-\-host \~}
.I HOST
] [
.RB {\~ \-p \~|\~ \-\-port \~}
.I PORT
]
.RB [\~ \-abBcdGHIKlLmMrRx
]
.RB [\~ \-g \fI\~SOURCE:FIRST\-LAST\fR
]
.RB [\~ \-i
.IR ATTR [, ATTR ]...\~]
.RB [\~ \-s
.IR SOURCE [, SOURCE ]...\~]
.RB [\~ \-T
.IR TYPE [, TYPE ]...\~]
.RB [\~ \-\-verbose \~]
.RB [\~ \-\-no\-recursion \~]
.I OBJECT

.B whois
.B \-q
.I KEYWORD

.B whois
.B \-t
.I TYPE

.B whois
.B \-v
.I TYPE

.B whois \-\-help

.B whois \-\-version

.SH DESCRIPTION
.B whois
searches for an object in a
.I RFC 3912
database.
.P
This version of the whois client tries to guess the right server to
ask for the specified object.
If no guess can be made it will connect to
.I whois.networksolutions.com
for NIC handles or
.I whois.arin.net
for IPv4 addresses and network names.
.SH OPTIONS
.TP 8
.B \-h
.IR HOST ,
.BI \-\-host= HOST
Connect to
.IR HOST .
.TP 8
.B \-H
Do not display the legal disclaimers that some registries like to show you.
.TP 8
.B \-p
.IR PORT ,
.BI \-\-port= PORT
Connect to
.IR PORT .
.TP 8
.B \-I
First query
.I whois.iana.org
and then follow its referral to the
whois server authoritative for that request.
This works for IP addresses,
AS numbers and domains.
.BR BEWARE :
this implies that the IANA server will receive your complete query.
.TP 8
.B \-\-no\-recursion
Disable recursion from registry to registrar servers.
.TP 8
.B \-\-verbose
Be verbose.
.TP 8
.B \-\-help
Display online help.
.TP 8
.B \-\-version
Display the program version.
.P
Other options are flags understood by
.I whois.ripe.net
and some other
RIPE-like servers:
.TP 8
.B \-a
Also search all the mirrored databases.
.TP 8
.B \-b
Return brief IP address ranges with abuse contact.
.TP 8
.B \-B
Disable objects filtering.
(Show the e-mail addresses.)
.TP 8
.B \-c
Return the smallest IP address range with a reference to an irt object.
.TP 8
.B \-d
Return the reverse DNS delegation object too.
.TP 8
.B \-g
.I SOURCE:FIRST\-LAST
Search updates from
.I SOURCE
database between
.I FIRST
and
.I LAST
update serial number.
It is useful to obtain Near Real Time Mirroring stream.
.TP 8
.B \-G
Disable grouping of associated objects.
.TP 8
.B \-i
.IR ATTR [, ATTR ]...
Inverse-search objects having associated attributes.
.I ATTR
is the attribute name, while the positional
.I OBJECT
argument is the attribute value.
.TP 8
.B \-K
Return primary key attributes only.
An exception is the
.I members
attribute of
.I set
objects, which is always returned.
Another exception are all
attributes of the objects
.IR organisation ,
.I person
and
.IR role ,
that are never returned.
.TP 8
.B \-l
Return the one level less specific object.
.TP 8
.B \-L
Return all levels of less specific objects.
.TP 8
.B \-m
Return all one level more specific objects.
.TP 8
.B \-M
Return all levels of more specific objects.
.TP 8
.B \-q
.I KEYWORD
Return information about the server.
.I KEYWORD
can be
.I version
for the server version,
.I sources
for the list of database sources or
.I types
for the list of supported object types.
.TP 8
.B \-r
Disable recursive lookups for contact information.
.TP 8
.B \-R
Disable following referrals and force showing the object from the local copy
in the server.
.TP 8
.B \-s
.IR SOURCE [, SOURCE ]...
Request the server to search for objects mirrored from
.IR SOURCE .
Sources are delimited by comma, and the order is significant.
Use the
.I \-q sources
parameter to obtain a list of valid sources.
.TP 8
.B \-t
.I TYPE
Return the template for a object of
.IR TYPE .
.TP 8
.B \-T
.IR TYPE [, TYPE ]...
Restrict the search to objects of
.IR TYPE .
Multiple types are separated by a comma.
.TP 8
.B \-v
.I TYPE
Return the verbose template for a object of
.IR TYPE .
.TP 8
.B \-x
Search for only exact match on network address prefix.
.SH NOTES
When querying the Verisign gTLDs
(e.g.\& \&.com, \&.net...\&)
thin registry servers
for a domain, the program will automatically prepend the
.I domain
keyword to only show domain records.  The
.I nameserver
or
.I registrar
keywords must be used to show other kinds of records.
.P
When querying
.I whois.arin.net
for IPv4 or IPv6 networks, the CIDR
netmask length will be automatically removed from the query string.
.P
When querying
.I whois.nic.ad.jp
for AS numbers, the program will automatically convert the request
in the appropriate format, inserting a space after the string
.IR AS .
.P
When querying
.I whois.denic.de
for domain names and no other
flags have been specified, the program will automatically add the flag
.IR "\-T dn" .
.P
When querying
.I whois.dk\-hostmaster.dk
for domain names and no other
flags have been specified, the program will automatically add the flag
.IR "\-\-show\-handles" .
.P
RIPE-specific command line options are ignored when querying non-RIPE
servers.
This may or may not be the behaviour intended by the user.
When using non-standard query parameters then the command line options
which are not to be interpreted by the client must follow the
.I \-\-
separator (which marks the beginning of the query string).
.P
If the
.I /etc/whois.conf
configuration file exists, it will be consulted
to find a server before applying the normal rules.
Each line of the
file should contain a regular expression to be matched against the query
text and the whois server to use, separated by white space.
IDN domains must use the ACE format.
.P
The whois protocol does not specify an encoding for characters which
cannot be represented by ASCII and implementations vary wildly.
If the program knows that a specific server uses a certain encoding,
if needed it will transcode the server output to the encoding specified
by the current system locale.
.P
Command line arguments will always be interpreted accordingly to the
current system locale and converted to the IDN ASCII Compatible Encoding.
.SH "FILES"
/etc/whois.conf
.SH "ENVIRONMENT"
.IP LANG
When querying
.I whois.nic.ad.jp
and
.I whois.jprs.jp
English text is requested unless the
.I LANG
or
.I LC_MESSAGES
environment variables specify a Japanese locale.
.IP "WHOIS_OPTIONS"
A list of options which will be evaluated before the ones specified on the
command line.
.IP "WHOIS_SERVER"
This server will be queried if the program cannot guess where some kind
of objects are located.
If the variable does not exist then
.I whois.arin.net
will be queried.
.SH "SEE ALSO"
.IR whois.conf (5).
.P
.IR "RFC 3912" :
WHOIS Protocol Specification.
.P
.IR "RIPE Database Query Reference Manual" :
.RI < https://www.ripe.net/\:data\-tools/\:support/\:documentation/\:ripe\-database\-query\-reference\-manual >
.SH BUGS
The program may have buffer overflows in the command line parser:
be sure to not pass untrusted data to it.
It should be rewritten to use a dynamic strings library.
.SH HISTORY
This program closely tracks the user interface of the whois client
developed at RIPE by Ambrose Magee and others on the base of the
original BSD client.
.SH AUTHOR
.B Whois
and this man page were written by Marco d'Itri
.RI < md@linux.it >
and are licensed under the terms of the GNU General Public License,
version 2 or later.
.\" SPDX-License-Identifier: GPL-2.0-or-later


debian/changelog
whois (5.5.23) unstable; urgency=medium

  * Updated the .sc, .新加坡 (.xn--yfro4i67o, Singapore) and .சிங்கப்பூர்
    (.xn--clchc0ea0b2g2a9gcd, Singapore) TLD servers.

 -- Marco d'Itri <md@linux.it>  Sat, 04 May 2024 03:13:22 +0200

whois (5.5.22) unstable; urgency=medium

  * Fixed a segmentation fault with --no-recursion.
  * Updated the .bm and .vi TLD servers.
  * Removed 4 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Mon, 01 Apr 2024 20:14:41 +0200

whois (5.5.21) unstable; urgency=medium

  * Updated the .cv and .sd TLD servers.
  * Remove 4 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Sat, 24 Feb 2024 17:22:22 +0100

whois (5.5.20) unstable; urgency=medium

  * Added the .gn TLD server.
  * Removed 6 new gTLDs which are no longer active.
  * Enabled getopt_long(3) support on Solaris.

 -- Marco d'Itri <md@linux.it>  Fri, 10 Nov 2023 15:26:02 +0100

whois (5.5.19) unstable; urgency=medium

  * Fixed english support for Japanese queries to not add again the /e
    argument if it had already been provided by the user. (Closes: #1050171)
  * Added the .ye and .বাংলা (.xn--54b7fta0cc, Bangladesh) TLD servers.
  * Updated the .ba, .bb, .dk, .es, .gt, .jo, .ml, .mo, .pa, .pn, .sv,
    .uy, .ﺍﻻﺭﺪﻧ (.xn--mgbayh7gpa, Jordan) and .澳門 (.xn--mix891f, Macao)
    TLD servers.
  * Upgraded the TLD URLs to HTTPS whenever possible.
  * Updated the charset for whois.jprs.jp.
  * Removed 3 new gTLDs which are no longer active.
  * Removed support for the obsolete as32 dot notation.

 -- Marco d'Itri <md@linux.it>  Sun, 08 Oct 2023 22:58:33 +0200

whois (5.5.18) unstable; urgency=medium

  [ Marco d'Itri ]
  * Updated the .ga TLD server. (Closes: #1037288)
  * Added new recovered IPv4 allocations.
  * Removed the delegation of 43.0.0.0/8 to JPNIC.
  * Removed 12 new gTLDs which are no longer active.
  * Improved the man page source, courtesy of Bjarni Ingi Gislason.
    (Closes: #1040613)

  [ Robert Scheck ]
  * Added the .edu.za SLD server.
  * Updated the .alt.za SLD server.
  * Added the -ru and -su NIC handles servers.

 -- Marco d'Itri <md@linux.it>  Sat, 22 Jul 2023 17:56:56 +0200

whois (5.5.17) unstable; urgency=medium

  [ Robert Scheck ]
  * Added the .cd TLD server.
  * Updated the -kg NIC handles server name.

  [ Marco d'Itri ]
  * Removed 2 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Wed, 03 May 2023 14:24:37 +0200

whois (5.5.16) unstable; urgency=medium

  * Add bash completion support, courtesy of Ville Skyttä.
  * Updated the .tr TLD server.
  * Removed support for -metu NIC handles.

 -- Marco d'Itri <md@linux.it>  Tue, 28 Feb 2023 00:58:22 +0100

whois (5.5.15) unstable; urgency=medium

  * Updated the .bd, .nz and .tv TLD servers.
  * Added the .llyw.cymru, .gov.scot and .gov.wales SLD servers.
  * Updated the .ac.uk and .gov.uk SLD servers.
  * Recursion has been enabled for whois.nic.tv.
  * Updated the list of new gTLDs with four generic TLDs assigned in
    October 2013 which were missing due to a bug.
  * Removed 4 new gTLDs which are no longer active.
  * Added the Georgian translation, contributed by Temuri Doghonadze.
  * Updated the Finnish translation, contributed by Lauri Nurmi.

 -- Marco d'Itri <md@linux.it>  Wed, 28 Dec 2022 20:44:11 +0100

whois (5.5.14) unstable; urgency=medium

  * Added the .bf and .sd TLD servers.
  * Removed the .gu TLD server.
  * Updated the .dm, .fj, .mt and .pk TLD servers.
  * Updated the charset for whois.nic.tr.
  * Updated the list of new gTLDs.
  * Removed whois.nic.fr from the list of RIPE-like servers, because it
    is not one anymore. (Closes: #1021110)
  * Renamed whois.arnes.si to whois.register.si in the list of RIPE-like
    servers.
  * Added the hiding string for whois.auda.org.au.
  * Fixed uclibc builds, because uclibc does not have NLS support.

 -- Marco d'Itri <md@linux.it>  Sun, 16 Oct 2022 22:16:34 +0200

whois (5.5.13) unstable; urgency=medium

  * Added the .sd TLD server.
  * Updated the list of new gTLDs.
  * Added the Turkish translation, contributed by Oğuz Ersen.

 -- Marco d'Itri <md@linux.it>  Fri, 08 Apr 2022 01:08:55 +0200

whois (5.5.12) unstable; urgency=medium

  * Updated the .pro TLD server, which was totally broken.
  * Fixed the detection of Japanese locales using $LC_MESSAGES.
  * Implemented providing partial salt strings to mkpasswd.
  * Removed 2 new gTLDs which are no longer active.
  * Updated one or more translations. (Closes: #1003597)
  * Enabled full hardening in debian/rules.

 -- Marco d'Itri <md@linux.it>  Wed, 23 Feb 2022 01:03:11 +0100

whois (5.5.11) unstable; urgency=medium

  * Implemented a --no-recursion command line option to disable recursion
    from registrar to registry servers.
  * Updated the .pro, .vu and .xxx TLD servers.
  * Updated the list of new gTLDs.
  * Removed 7 new gTLDs which are no longer active.
  * Updated make_version_h.pl to support Ubuntu no-change uploads,
    contributed by Matthias Klose. (Closes: #995873)

 -- Marco d'Itri <md@linux.it>  Mon, 03 Jan 2022 18:18:36 +0100

whois (5.5.10) unstable; urgency=medium

  * Updated the .lb TLD server.
  * Removed 5 new gTLDs which are no longer active.
  * Updated the charset for whois.lacnic.net, whois.ax, whois.cira.ca
    and whois.dns.pt.
  * Fixed reporting an older version number when using --version.

 -- Marco d'Itri <md@linux.it>  Sun, 06 Jun 2021 19:54:13 +0200

whois (5.5.9) unstable; urgency=medium

  * Updated the .ga TLD server.
  * Removed the .cd and cf TLD servers.
  * Removed 72 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Sun, 28 Mar 2021 00:38:20 +0100

whois (5.5.8) unstable; urgency=medium

  * Added the .xn--4dbrk0ce (.ישראל, Israel) TLD server.
  * Updated the .ie and .az TLD servers.
  * Updated the list of new gTLDs.

 -- Marco d'Itri <md@linux.it>  Tue, 16 Feb 2021 01:53:57 +0100

whois (5.5.7) unstable; urgency=medium

  * Implemented support for CIDR queries to whois.arin.net.

whois.bash
_whois_query() {
	"$1" -q "$2" 2>/dev/null | while read -r item _; do
		[[ $item == %* ]] && continue
		printf "%s\n" "${item%%:*}"
	done
}

_whois_hosts() {
	# _known_hosts_real from github.com/scop/bash-completion if available
	if declare -f _known_hosts_real &>/dev/null; then
		_known_hosts_real -- "$1"
		return 0
	fi
	COMPREPLY=($(compgen -A hostname -- "$1"))
}

_whois() {

	case $3 in
	--help | --version | -p | --port | -i)
		return 0
		;;
	-h | --host)
		_whois_hosts "$2"
		return 0
		;;
	-T | -t | -v)
		[[ ${_whois_types-} ]] ||
			_whois_types=" $(_whois_query "$1" types)"
		COMPREPLY=($(compgen -W '$_whois_types' -- "$2"))
		return 0
		;;
	-s | -g)
		[[ ${_whois_sources-} ]] ||
			_whois_sources=" $(_whois_query "$1" sources)"
		COMPREPLY=($(compgen -W '$_whois_sources' -- "$2"))
		if [[ $3 == -g ]]; then
			[[ ${#COMPREPLY[*]} -eq 1 ]] && COMPREPLY[0]+=:
			compopt -o nospace
		fi
		return 0
		;;
	-q)
		COMPREPLY=($(compgen -W 'version sources types' -- "$2"))
		return 0
		;;
	esac

	if [[ $2 == -* ]]; then
		COMPREPLY=($(compgen -W '
			-h --host
			-p --port
			-I
			-H
			--verbose
			--no-recursion
			--help
			--version
			-l
			-L
			-m
			-M
			-c
			-x
			-b
			-B
			-G
			-d
			-i
			-T
			-K
			-r
			-R
			-a
			-s
			-g
			-t
			-v
			-q
		' -- "$2"))
		return 0
	fi

	_whois_hosts "$2"

} && complete -F _whois whois
